feat: add `this.Notify` and `this.Provide` analyzers and code fixes

ptlthg opened the pull request:
This PR adds code analyzers and fixes for common mistakes when using AutoInject
- Forgetting to call `this.Notify`
- Forgetting to call `this.Provide`

This is my first time actually making something functional when it comes to Roslyn tools.

https://github.com/user-attachments/assets/4022761a-7530-4fab-8216-ab25f98501e9

## Analyzers

Three diagnostics are added:

- **`AUTO_INJECT_001` Missing "_Notification" method override**
Emitted when `_Notification` isn't overridden in any class that has any of the 6 AutoInject meta interfaces. 
- **`AUTO_INJECT_002` Missing "this.Notify(what)" method call**
Emitted when there is a `_Notification` method, but `this.Notify` isn't called within it.
- **`AUTO_INJECT_003` Missing "this.Provide()" call in provider class**
Emitted when a class has `IAutoNode` or `IProvider`, implements `IProvide<T>`, and doesn't have `this.Provide()` called anywhere in the class.

## Code Fixes

To help fix those issues, a handful of code fixes were added:

- **Add "this.Notify(what);" to existing \"_Notification\" override**
Does what it says really, it will also change expression syntax to block syntax if needed.
- **Add \"public override void _Notification(int what) => this.Notify(what);\" method**
Probably will be the most used, it directly adds that line to the top of the class.

Then for missing `this.Provide` calls we have a lot of options.

- `(if Setup() exists)` **Add "this.Provide();" to existing "Setup()" method**
	- `(else)` **Create "Setup()" method that calls "this.Provide();"**
- `(if OnReady() exists)` **Add "this.Provide();" to existing "OnReady()" method**
	- `(else)` **Create "OnReady()" method that calls "this.Provide();"**
- `(if _Ready()` exists) **Add "this.Provide();" to existing "_Ready()" method**
	- ~~I figured that a create `_Ready()` option wasn't needed because that's not really in the domain of AutoInject~~ **Edit:** Added a create option for this too
	
## Tests

I don't know how unit tests might be added for these, ~~feel free to add them for me~~. Just from testing it while making it, everything seems to work great, but it should be tested more.



## Issues

I don't know how to get this published as a NuGet package, I think I changed the appropriate values in the csproj but I don't know. Any help there would be appreciated.

The code fixes are in the same project as the analyzers, this might be a mistake? I had to silence [`RS1038`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/rules/RS1038.md) on the analyzers because of this, but I wasn't sure if I should split the code fixes into a new project or not, as it sounds like this warning can be worked around or doesn't apply to this according to https://github.com/dotnet/roslyn-analyzers/issues/7438

### Future

I want to add analyzers and code fixes for using `OnProcess` and `OnPhysicsProcess` without calling `SetPhysicsProcess(true);` or `SetProcess(true);`. Probably should go in a new PR though.

ptlthg commented on Chickensoft.AutoInject.Analyzers/src/fixes/AutoInjectProvideFixProvider.cs:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    // If they have _Ready(), suggest adding this.Provide() at the end of it.
+    if (hasOnReadyOverride) {
```
I mentioned this in the PR description, but my rationale here was that _Ready() isn't an autoinject method so we shouldn't suggest making it. Could definitely still add it anyways though.

ptlthg replied:
_Notification is different because it's actually required, but I can add a create _Ready override option.

ptlthg replied:
Added the option to create _Ready()

blewis-web commented:
I don't know anything about how to make Analyzers, but I found the unit tests for `dotnet/roslyn-analyzers/` in case that is helpful (linked to a simple test):
https://github.com/dotnet/roslyn-analyzers/blob/ab7c0aff23c675c8fde519abe05f2cb884eec450/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs#L119

ptlthg commented on Chickensoft.AutoInject.Analyzers/src/AutoInjectProvideAnalyzer.cs:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    // Check that Meta attribute has an AutoInject Provider type (ex: [Meta(typeof(IAutoNode))])
+    var attributes = classDeclaration.AttributeLists.SelectMany(list => list.Attributes
+      ).Where(attribute => attribute.Name.ToString() == Constants.META_ATTRIBUTE_NAME
+         && attribute.ArgumentList?.Arguments.Any(arg =>
+           arg.Expression is TypeOfExpressionSyntax { Type: IdentifierNameSyntax identifierName } &&
+           Constants.ProviderMetaNames.Contains(identifierName.Identifier.ValueText)
+         ) == true
+      )
```
Same as I commented above, the equals true is needed since the left side is nullable

ptlthg commented:
> I found the unit tests for `dotnet/roslyn-analyzers/` in case that is helpful

I did already see some examples of how unit testing is possible, I'm just not sure how/where it should be done for this package, really. I'm not sure if this something that would work in the existing tests project or if a new one should be made, or really anything along those lines. As far as I know, the existing test project has some weird things happening with the build, so I don't want to touch that without knowing what's up.

I probably need @jolexxa or @wlsnmrk to help out with where the analyzer tests should go/how it should be set up.

ptlthg commented:
I cleaned up the provide code fix file a lot and I added "create _Ready() with this.Provide" as another option too. I just ran the built-in Rider formatter on that file also, I think it worked okay with the code styles so hopefully we're good on style there

wlsnmrk commented:
> I did already see some examples of how unit testing is possible, I'm just not sure how/where it should be done for this package, really. I'm not sure if this something that would work in the existing tests project or if a new one should be made, or really anything along those lines. As far as I know, the existing test project has some weird things happening with the build, so I don't want to touch that without knowing what's up.
> 
> I probably need @jolexxa or @wlsnmrk to help out with where the analyzer tests should go/how it should be set up.

I would put them in a separate test project, not in the main AutoInject.Tests.

There's [a Visual Studio template](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix) for creating analyzer solutions including test projects, which appears to include some helpful infrastructure for running tests on the analyzers. I haven't had any luck tracking down a general template for use with `dotnet new`, but it seems most of the template code is [here](https://github.com/dotnet/roslyn-sdk/tree/main/src/VisualStudio.Roslyn.SDK/Roslyn.SDK/ProjectTemplates/CSharp/Diagnostic/Test/Verifiers), so you might be able to copy those C# classes over to a standard test project and go from there.

wlsnmrk commented:
The .NET Roslyn analyzers [seem to put analyzers and fixers in the same project](https://github.com/dotnet/roslyn-analyzers/tree/main/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines), so I think that's fine as a design decision.

I'm not too worried about the RS1038. In fact, VSCode flags the suppression pragma as removable and doesn't even show the warning if I take out the pragma.

I've just done some manual testing in one of my projects using AutoInject and everything seems to work okay.

ptlthg commented:
> I'm not too worried about the RS1038. In fact, VSCode flags the suppression pragma as removable and doesn't even show the warning if I take out the pragma.

Good to know, thanks. I also tried this in my existing game project and got no false positives and everything seemed to work well there too

jolexxa commented:
@ptlthg Would you mind running this on the Game Demo and seeing if it produces any false positives?

ptlthg commented:
> @ptlthg Would you mind running this on the Game Demo and seeing if it produces any false positives?

Did that just now, I don't get any false positives but I did notice that it doesn't pick up that you need `this.Provide();` in a situation like this:

```csharp
public interface IApp : ICanvasLayer, IProvide<IAppRepo>;

[Meta(typeof(IAutoNode))]
public partial class App : CanvasLayer, IApp {
```

The analyzer just checks for if IProvide is directly implemented on the class and doesn't check IApp for it. Maybe it should check if you have a `IProvide<T>.Value()` method implementation instead?

jolexxa commented:
Ooh, interesting catch.

> Maybe it should check if you have a IProvide<T>.Value() method implementation instead?

That sounds like the most performant way to make this check.

ptlthg commented:
> That sounds like the most performant way to make this check.

I ended up using the semantic model to get all interfaces to fix this instead, it only took a couple lines of change and from what I read it doesn't sound like there would be a large performance impact by doing this. But let me know if you have concerns about it.

wlsnmrk commented:
Tried this out on a personal project and the GameDemo, and I don't see any noticeable performance impact in use.

I do still need to test the nuget packaging and make sure it's working as expected.

wlsnmrk commented:
Nuget package pushed to a local source seems to work fine 👍 I've added the analyzer project to the release workflow - would appreciate a quick glance from @jolexxa to verify that it seems okay.
